edge/yaml/docker_compose.go

```diff
@@ -57,30 +51,77 @@ func addEnvsForSpecificService(fileContent, serviceName string, envs map[string]
 		return "", errors.New("Failed to validate the compose file content")
 	}
 
+	// Extract registry server url from compose object
 	service, ok := compose.Services[serviceName]
 	if !ok {
-		return "", errors.Wrap(err, fmt.Sprintf("Cannot find the service: %s", serviceName))
+		return "", fmt.Errorf("failed to find the service: %s", serviceName)
 	}
 
+	serverUrl, err := extractRegistryServerUrl(service.Image)
+	if err != nil {
+		return "", err
+	}
+
+	// Generate envs
+	envs := make(map[string]string)
+	if y.awsConfig != nil {
+		log.Info().Msg("using local AWS config for credential lookup")
+
+		// Exchange ECR credential with ECR certificate
+		c, err := aws.DoAWSIAMRolesAnywhereAuthAndGetECRCredentials(serverUrl, y.awsConfig)
+		if err != nil {
+			return "", err
+		}
+
+		envs["REGISTRY_USED"] = "1"
+		// hardcode username for aws ecr registry
+		// @https://docs.aws.amazon.com/cli/latest/reference/ecr/get-login-password.html#examples
+		envs["REGISTRY_USERNAME"] = "AWS"
+		envs["REGISTRY_PASSWORD"] = c.Secret
+		envs["REGISTRY_PASSWORD"] = c.Secret
+
+	} else if len(y.RegistryCredentials) > 0 {
+		log.Info().Msg("using private registry credential")
+
+		for _, cred := range y.RegistryCredentials {
+			if serverUrl == cred.ServerURL {
+				log.Info().Str("registry server url", cred.ServerURL)
+
+				envs["REGISTRY_USED"] = "1"
+				envs["REGISTRY_USERNAME"] = cred.Username
+				envs["REGISTRY_PASSWORD"] = cred.Secret
+				break
+			}
+		}
+	}
+
+	return updateServiceWithEnv(compose, serviceName, envs)
```

Comment:
maybe we should think builder pattern. so we call this:
```
ym := yaml.NewDockerComposeYaml(....); // here we already have parsed ym as `ym.parsed` or `ym.data`
ym.WithCredentails("updater")
ym.WithEnv("updater", envs)
newFile := ym.String()
```

wdyt?

Reply:
If we're only working with Docker Compose YAML files, I don't think we need to create a pattern right away. Alternatively, we could consider developing a pattern that covers both Docker Compose YAML and K8s YAML in a separate project to make the workflow more clear and organized.